### PR TITLE
feat: add withdraw logic to Collateral

### PR DESCRIPTION
**Pull request by ramenforbreakfast**

### CODE REVIEW

**Review by XavierEkkel — COMMENTED** [review]
please adjust this PR with similar comments from https://github.com/prepo-io/prepo-monorepo/pull/496, and then re-request review!

**Comment by ramenforbreakfast** [comment]
I have updated this PR with the following changes:
- Add ReentrancyGuard
- Added tests for different base tokens
- Changed revert case to follow the same format as `deposit` in the prior PR
- Changed event to follow the `Deposit` format

**Comment by ramenforbreakfast** [comment]
Updated this PR once again, introducing changes that remove transfers of fee to treasury and instead provide approval to the `withdrawHook`. Also applied various minor nits that were mentioned in PR 503 as well.

File: apps/smart-contracts/core/contracts/Collateral.sol [hunk 1]
```diff
@@ -71,7 +71,27 @@ contract Collateral is
     emit Deposit(msg.sender, _amountAfterFee, _fee);
   }
 
-  function withdraw(uint256 _amount) external override {}
+  function withdraw(uint256 _amount) external override nonReentrant {
+    uint256 _baseTokenAmount = (_amount * _baseTokenDenominator) / 1e18;
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/contracts/Collateral.sol:75]: add comment above this line:
`Converts amount from collateral token units to base token units.`

File: apps/smart-contracts/core/test/Collateral.test.ts [hunk 2]
```diff
@@ -834,4 +848,275 @@ describe('=> Collateral', () => {
...
+    beforeEach(async () => {
+      await setupCollateral()
+      await baseToken.mint(user1.address, parseUnits('1', USDC_DECIMALS))
+      await baseToken.connect(user1).approve(collateral.address, parseUnits('1', USDC_DECIMALS))
+      await collateral.connect(user1).deposit(parseUnits('1', USDC_DECIMALS))
+      await collateral.connect(deployer).setWithdrawFee(TEST_WITHDRAW_FEE)
+      await collateral.connect(deployer).setWithdrawHook(withdrawHook.address)
+    })
+
+    it('reverts if withdrawal 0 and fee rate 0', async () => {
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/Collateral.test.ts:862]: same comments about tests here as in the deposit logic PR

File: apps/smart-contracts/core/test/Collateral.test.ts [hunk 3]
```diff
@@ -834,4 +848,275 @@ describe('=> Collateral', () => {
...
+      const amountToWithdraw = (await collateral.balanceOf(user1.address)).add(1)
+      expect(amountToWithdraw).to.be.gt(0)
+
+      await expect(collateral.connect(user1).withdraw(amountToWithdraw)).revertedWith(
+        'ERC20: burn amount exceeds balance'
+      )
+    })
+
+    it('reverts if hook reverts', async () => {
+      // Still providing valid inputs to ensure withdrawals are only reverting due to smock
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/Collateral.test.ts:910]: cant you also check that the smock is called?

File: apps/smart-contracts/core/test/Collateral.test.ts [hunk 4]
```diff
@@ -834,4 +848,275 @@ describe('=> Collateral', () => {
...
+      await collateral.connect(user1).deposit(parseEther('1'))
+      const amountToWithdraw = await collateral.balanceOf(user1.address)
+      expect(amountToWithdraw).to.be.gt(0)
+      const expectedBT = amountToWithdraw
+      const fee = expectedBT.mul(await collateral.getWithdrawFee()).div(FEE_DENOMINATOR)
+      expect(await baseToken.allowance(collateral.address, withdrawHook.address)).to.be.eq(0)
+
+      await collateral.connect(user1).withdraw(amountToWithdraw)
+
+      expect(await baseToken.allowance(collateral.address, withdrawHook.address)).to.be.eq(fee)
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/Collateral.test.ts:976]: add assertion that fee balance stays in the collateral contract

**Review by XavierEkkel — APPROVED** [review]
LGTM after comments

**Pull request merged**